### ISSUE TITLE
Reconcile timeout syntax between config file and HTTPProxy API

### ISSUE THREAD

**Issue by skriss** [projectcontour/contour#2697]
There are some inconsistencies in the syntax for timeouts between the `contour serve` config file and the HTTPProxy API.

In the HTTPProxy API:
- `infinity` is a valid timeout value and results in a `0s` being explicitly passed to Envoy (which Envoy generally interprets as disabling the timeout in question)
- `0` is interpreted as if the timeout were not specified, and results in no value being passed to Envoy, so Envoy's default behavior applies
- an unparseable string is treated the same as "infinity"

In the config file:
- `infinity` is *not* a valid timeout value since we're YAML parsing into `time.Duration` fields.
- `0s` is interpreted as "disable this timeout", so *usually* results in `0s` being explicitly passed to Envoy to disable the timeout (`max_connection_duration` is an exception to this, though, since it needs to be omitted to be disabled).
- if the field is not specified, the Contour defaults apply, which may or may not be the same as the Envoy defaults
    - there is no true fallback to Envoy default behavior; however, per parent bullet, for some timeout fields Contour explicitly passes the current Envoy default as the timeout value if it wasn't specified in the config file (stream_idle, drain).
- an unparseable string is treated as if the field were not specified

I think for UX consistency, we should adopt the current HTTPProxy syntax in both places, so:
- change the config file timeout fields to strings
- allow `infinity`, which means "disable this timeout"
- not specifying the field at all falls back to either a Contour default (e.g. connection_idle), or the Envoy default behavior (others)
- `0` would also be interpreted as if the field were not specified
- an unparseable string is treated the same as "infinity"
    - I have a weak preference for changing this behavior, so it uses the Contour/Envoy default instead, but we can do that separately if we want to

----

`drain_timeout` is weird, because `infinity` doesn't make sense for disabling it -- rather, setting it to 0 makes sense, since it's really a grace period, not a timeout (specifically, it's the time between sending a "warning" GOAWAY frame and sending a final GOAWAY frame).

Perhaps our magic string should be "disabled" instead of "infinity"?

**Comment by youngnick** [comment]
I think that adding `disabled` is probably good, it's certainly a clearer way to indicate that the timeout is disabled. I love the idea of making things consistent everywhere though, and the PR.

**Comment by jpeach** [comment]
> change the config file timeout fields to strings

👍 

> allow infinity, which means "disable this timeout"

Just to quibble with this a bit, disabling a timeout means that the operation doesn't time out, so that's equivalent to timing out after "infinity". When you look at it this way, a drain timeout of "infinity" does make sense (it means you can wait forever for the drain to occur).

> not specifying the field at all falls back to either a Contour default

Yup, just need to ensure the API reference has useful information about the default.

> 0 would also be interpreted as if the field were not specified

If timeouts are always strings, then if the field is not specified it's value is "", so I don't think we need to do this. IIRC the 0 is only magic because Contour doesn't have a structured internal type for representing timeouts.

> an unparseable string is treated the same as "infinity"

This should be a validation error. For config, I would make it a hard error, for HTTPProxy resources, I'd also be sympathetic to exposing the validation error as a warning on the new status condition. We can use `+kubebuilder:validation:Pattern` to validate timeout string syntax.

**Comment by youngnick** [comment]
I agree with James here, about everything he mentioned, and would like to strongly recommend we put anything unparseable in a HTTPProxy into either a validation error or warning in the new status condition. I'm just wrapping up the docstrings for adding that to HTTPProxy today.

**Comment by skriss** [comment]
> > allow infinity, which means "disable this timeout"

> Just to quibble with this a bit, disabling a timeout means that the operation doesn't time out, so that's equivalent to timing out after "infinity". When you look at it this way, a drain timeout of "infinity" does make sense (it means you can wait forever for the drain to occur).

I agree that "infinity" is equally valid as "disabled" in most places. The problem with drain_timeout is that disabling it, by passing a 0, results in Envoy sending the second GOAWAY frame immediately after the first, rather than waiting an infinite amount of time. It's really not a timeout in the sense of "wait up to a certain amount of time for something to happen", it's just a fixed grace period/duration. If you read the [Envoy docs](https://www.envoyproxy.io/docs/envoy/latest/api-v2/config/filter/network/http_connection_manager/v2/http_connection_manager.proto.html) that's how they refer to it, despite the name.

> > 0 would also be interpreted as if the field were not specified

 > If timeouts are always strings, then if the field is not specified it's value is "", so I don't think we need to do this. IIRC the 0 is only magic because Contour doesn't have a structured internal type for representing timeouts.

I'm happy to make this change as I do think the current behavior is a little strange. Note that this will change API semantics though - are we comfortable with that? If so, I'll tackle this in a separate PR.

> > an unparseable string is treated the same as "infinity"

> This should be a validation error. For config, I would make it a hard error, for HTTPProxy resources, I'd also be sympathetic to exposing the validation error as a warning on the new status condition. We can use +kubebuilder:validation:Pattern to validate timeout string syntax.

This makes sense as well, though the same caveats around changing API behavior apply. I'll tackle this in a separate PR if we want to go ahead.

**Comment by jpeach** [comment]
> > > allow infinity, which means "disable this timeout"
> 
> > Just to quibble with this a bit, disabling a timeout means that the operation doesn't time out, so that's equivalent to timing out after "infinity". When you look at it this way, a drain timeout of "infinity" does make sense (it means you can wait forever for the drain to occur).
> 
> I agree that "infinity" is equally valid as "disabled" in most places. The problem with drain_timeout is that disabling it, by passing a 0, results in Envoy sending the second GOAWAY frame immediately after the first, rather than waiting an infinite amount of time. It's really not a timeout in the sense of "wait up to a certain amount of time for something to happen", it's just a fixed grace period/duration. If you read the [Envoy docs](https://www.envoyproxy.io/docs/envoy/latest/api-v2/config/filter/network/http_connection_manager/v2/http_connection_manager.proto.html) that's how they refer to it, despite the name.

In this case, I'd argue that it is a timeout in the sense of "wait up to a certain amount of time before doing something". Adding the explicit `disabled` representation would let us correctly handle this (by contextually setting an infinite duration), but that would give a nonsensical result (the session would never drain) which we probably ought to disallow in validation.

In this particular case, it might make sense to expose `drain_timeout` with a more intuitive name, maybe `SessionShutdownGracePeriod`.

> > > 0 would also be interpreted as if the field were not specified
> 
> > If timeouts are always strings, then if the field is not specified it's value is "", so I don't think we need to do this. IIRC the 0 is only magic because Contour doesn't have a structured internal type for representing timeouts.
> 
> I'm happy to make this change as I do think the current behavior is a little strange. Note that this will change API semantics though - are we comfortable with that? If so, I'll tackle this in a separate PR.

Can you explain the semantic change to me? The way I was thinking about it was that "" has always meant "be default"

> > > an unparseable string is treated the same as "infinity"
> 
> > This should be a validation error. For config, I would make it a hard error, for HTTPProxy resources, I'd also be sympathetic to exposing the validation error as a warning on the new status condition. We can use +kubebuilder:validation:Pattern to validate timeout string syntax.
> 
> This makes sense as well, though the same caveats around changing API behavior apply. I'll tackle this in a separate PR if we want to go ahead.

I feel OK with this since it defines API behavior that is currently undefined, but I'm interested to hear other perspectives.

**Comment by skriss** [comment]
> > > > allow infinity, which means "disable this timeout"
> > 
> > 
> > > Just to quibble with this a bit, disabling a timeout means that the operation doesn't time out, so that's equivalent to timing out after "infinity". When you look at it this way, a drain timeout of "infinity" does make sense (it means you can wait forever for the drain to occur).
> > 
> > 
> > I agree that "infinity" is equally valid as "disabled" in most places. The problem with drain_timeout is that disabling it, by passing a 0, results in Envoy sending the second GOAWAY frame immediately after the first, rather than waiting an infinite amount of time. It's really not a timeout in the sense of "wait up to a certain amount of time for something to happen", it's just a fixed grace period/duration. If you read the [Envoy docs](https://www.envoyproxy.io/docs/envoy/latest/api-v2/config/filter/network/http_connection_manager/v2/http_connection_manager.proto.html) that's how they refer to it, despite the name.
> 
> In this case, I'd argue that it is a timeout in the sense of "wait up to a certain amount of time before doing something". Adding the explicit `disabled` representation would let us correctly handle this (by contextually setting an infinite duration), but that would give a nonsensical result (the session would never drain) which we probably ought to disallow in validation.

OK, so if we un-special-case 0 per the below discussion, then a user can specify a 0 drain_timeout if that's what they want. Additionally, we can explicitly disallow `infinity` for this field, since it's nonsensical. If we do those things, then we don't need to add a `disabled` keyword, since `infinity` works fine for other fields, and 0 works for this field.

I *think* that should be workable for you, but let me know if I'm missing something.

> In this particular case, it might make sense to expose `drain_timeout` with a more intuitive name, maybe `SessionShutdownGracePeriod`.

Agree with this.

> > > > 0 would also be interpreted as if the field were not specified
> > 
> > 
> > > If timeouts are always strings, then if the field is not specified it's value is "", so I don't think we need to do this. IIRC the 0 is only magic because Contour doesn't have a structured internal type for representing timeouts.
> > 
> > 
> > I'm happy to make this change as I do think the current behavior is a little strange. Note that this will change API semantics though - are we comfortable with that? If so, I'll tackle this in a separate PR.
> 
> Can you explain the semantic change to me? The way I was thinking about it was that "" has always meant "be default"

Yep, semantics of "" wouldn't change. Semantics of an explicit 0 would change for the API, though. Currently, it's interpreted the same as "" (i.e. use the default), but if we un-special-case it and treat it like any other numerical value, then it'll usually mean "disable the timeout" (i.e. same as "infinity") since that's how Envoy interprets it.

> > > > an unparseable string is treated the same as "infinity"
> > 
> > 
> > > This should be a validation error. For config, I would make it a hard error, for HTTPProxy resources, I'd also be sympathetic to exposing the validation error as a warning on the new status condition. We can use +kubebuilder:validation:Pattern to validate timeout string syntax.
> > 
> > 
> > This makes sense as well, though the same caveats around changing API behavior apply. I'll tackle this in a separate PR if we want to go ahead.
> 
> I feel OK with this since it defines API behavior that is currently undefined, but I'm interested to hear other perspectives.

I also think it's OK, just wanted to double-check.

**Comment by jpeach** [comment]
> > > > > 0 would also be interpreted as if the field were not specified
> > > 
> > > 
> > > > If timeouts are always strings, then if the field is not specified it's value is "", so I don't think we need to do this. IIRC the 0 is only magic because Contour doesn't have a structured internal type for representing timeouts.
> > > 
> > > 
> > > I'm happy to make this change as I do think the current behavior is a little strange. Note that this will change API semantics though - are we comfortable with that? If so, I'll tackle this in a separate PR.
> > 
> > 
> > Can you explain the semantic change to me? The way I was thinking about it was that "" has always meant "be default"
> 
> Yep, semantics of "" wouldn't change. Semantics of an explicit 0 would change for the API, though. Currently, it's interpreted the same as "" (i.e. use the default), but if we un-special-case it and treat it like any other numerical value, then it'll usually mean "disable the timeout" (i.e. same as "infinity") since that's how Envoy interprets it.

Thanks for clarifying, @skriss I lost track of the thread a bit :)

So, I think that since we have explicit timeout string semantics now, we should not combine these with implicit Envoy interpretations. I agree with your original proposal to treat "0s" and "" the same. Since the config file parsing has treated "0s" as "infinity", I also agree we need to think the compatibility issue through for that.

**Comment by youngnick** [comment]
So here's what I think we should do for the config file timeout, since there's only one (This is covered by #2690):
- We add a new `requestTimeout` field to the `timeouts` stanza. This has the same behavior as all the other timeouts (that is, `0s` means "use the default", not "disable"). It should be clearly called out in the documentation for this field that the semantics have changed from the standalone field.
- We deprecate the current standalone request timeout field, with a remove in n versions (where n is probaly 3-6). Using this field will cause Contour to log at warning level, with a pointer to the new field's documentation maybe?

I don't think there are any other standalone timeouts in the config file at the moment, if so, we should treat them the same.

So, to summarise what I can see here:
- all timeouts in the new timeouts config file struct should use the standard timeout parsing, the same as HTTPProxy.
- request timeout should be moved in there asap (that is, #2690)

Have we done a release with the timeouts config in the config file yet? If not, this is our one chance to get this right before next week.

**Comment by skriss** [comment]
considering this closed by #2698 and #2721. I opened #2728 to track handling parse errors across the board (config file, HTTPProxy, Ingress).
